### ISSUE TITLE
Custom date types & `AbstractConnection.current{Date,DateTime,Time,Timestamp}()`

### ISSUE THREAD

**Issue by beverloo** [juanluispaz/ts-sql-query#122]
Thank you for introducing the custom date and time types in 1.59, that removed a whole bunch of complexity for me as I switched to [`Temporal`](https://tc39.es/proposal-temporal/docs/), the upcoming JavaScript `Date` replacement 😃 

One corner case is that the `AbstractConnection.current{Date,DateTime,Time,Timestamp}()` methods don't have a TypeScript signature that allows them to be used, requiring them [to be overridden](https://github.com/beverloo/volunteer-manager/blob/main/app/lib/database/Connection.ts#L50) adding the more specific type. They work fine with that addition.

I'm not entirely sure whether this is fixable with the current signatures—the specific type used for custom dates and times may be specific to a particular column, whereas the methods on `AbstractConnection` are global to the database.

One idea could be to pass the relevant column as an argument to e.g. `currentDateTime` and derive the type from that, but that feels a bit hackish:

```javascript
await db.update(tTable)
  .set({
    updated: db.currentDateTime(tTable.updated),
  })
  .where(tTable.id.equals(1))
  .executeUpdate();
```

This isn't a major problem, but I figured I'd flag it!

**Comment by juanluispaz** [comment]
Hi!

Wonderful project you have, and thank you for the words in the readme.

I really want Temporal API rolled out in JavaScript, which will make a lot of things easier.

May I suggest defining your own [reusable SQL Fragment](https://ts-sql-query.readthedocs.io/en/stable/queries/sql-fragments/#select-with-custom-reusable-sql-fragment) with a function that returns the current date-time in the desired type. You can define your DBConnection class as:

```ts
class DBConnection extends MariaDBConnection<'DBConnection'> { 

    currentZonedDateTime = this.buildFragmentWithArgs().as(() => {
        return this.fragmentWithType<ZonedDateTime>('customLocalDateTime', 'ZonedDateTime', 'required')
            .sql`current_timestamp`
    })

    currentPainDate = this.buildFragmentWithArgs().as(() => {
        return this.fragmentWithType<PlainDate>('customLocalDate', 'PainDate', 'required')
            .sql`current_date`
    })

    zonedDateTimeAsString = this.buildFragmentWithArgs(
        this.arg<ZonedDateTime>('customLocalDateTime', 'ZonedDateTime', 'required')
    ).as((value) => {
        return this.fragmentWithType('string', 'required')
            .sql`date_format(${value}, "%Y-%m-%dT%TZ[UTC]")`
    })

    plainDateAsString = this.buildFragmentWithArgs(
        this.arg<PlainDate>('customLocalDate', 'PainDate', 'required')
    ).as((value) => {
        return this.fragmentWithType('string', 'required')
            .sql`date_format(${value}, "%Y-%m-%d")`
    })

    protected transformValueFromDB(value: unknown, type: string) {
        if (type === 'ZonedDateTime' && value) {
            // TODO: validate returned value from DB
            return Temporal.ZonedDateTime.from(value)
        } else if (type === 'PlainDate' && value) {
            // TODO: validate returned value from DB
            return Temporal.PlainDate.from(value)
        }
        return super.transformValueFromDB(value, type)
    }
    protected transformValueToDB(value: unknown, type: string): unknown {
        if (type === 'ZonedDateTime' && value) {
            if (!(value instanceof Temporal.ZonedDateTime)) {
                throw new Error('Invalid ZonedDateTime value to send to the db: ' + value);
            }

            return formatDate(value.withTimeZone('UTC'), 'YYYY-MM-DD HH:mm:ss')
        } else if (type === 'PainDate' && value) {
            if (!(value instanceof Temporal.ZonedDateTime)) {
                throw new Error('Invalid ZonedDateTime value to send to the db: ' + value)
            }

            return formatDate(value, 'YYYY-MM-DD')
        }
        return super.transformValueToDB(value, type)
    }
}
```

Then, you can define your tables as:

```ts
const tMyTable = new class tMyTable extends Table<DBConnection, 'TMyTable'> {
    id = this.autogeneratedPrimaryKey('id', 'int')
    myDate = this.column<PlainDate>('my_date_column', 'customLocalDate', 'PainDate')
    myDateTime = this.column<ZonedDateTime>('my_datetime_column', 'customLocalDate', 'ZonedDateTime')
    constructor() {
        super('my_table')
    }
}()
```

You can use in your queries as:

```ts
    let dateValue: PlainDate = ...
    let dateTimeValue: ZonedDateTime = ...

    const insertedId = await connection.insertInto(tMyTable)
        .set({
            myDate: dateValue,
            myDateTime: dateTimeValue
        })
        .returningLastInsertedId()
        .executeInsert()
```

```ts
    const insertedId = await connection.insertInto(tMyTable)
        .set({
            myDate: connection.currentPainDate(),
            myDateTime: connection.currentZonedDateTime()
        })
        .returningLastInsertedId()
        .executeInsert()
```

```ts
    const selectedData = await connection.selectFrom(tMyTable)
        .select({
            dateValue: tMyTable.myDate,
            dateTimeValue: tMyTable.myDateTime,
            dateValueAsString: connection.plainDateAsString(tMyTable.myDate),
            dateTimeValueAsString: connection.zonedDateTimeAsString(tMyTable.myDate),
        })
```

I put the type adaptations in the DB connection. You can use a separated type adapter, but you will need to specify everywhere. Perhaps defining it globally is easier, and you don't forget to put it anywhere.

Something caught my attention in your code: the file [DateTimeStringColumns](https://github.com/beverloo/volunteer-manager/blob/main/app/lib/database/DateTimeStringColumns.ts). In a quick search, I didn't find where you are using it.

I would prefer, if possible, you not use the internal types in your project because over the time that types changes; I can add additional features to ts-sql-query if you identify something that requires the internal types. Each project has its own needs, and identifying it allows us to grow ts-sql-query.

Let me know if this works for you, and don't hesitate to contact me if you need to ask a question or request a new feature.

**Comment by beverloo** [comment]
Many thanks for your recommendations, @juanluispaz!

I completely missed the `buildFragmentWithArgs()` & `arg()` methods, that's much cleaner than the attempts I've previously had indeed. It was [super easy](https://github.com/beverloo/volunteer-manager/commit/c3db2474e52d1f9c435b461be0055c79bf0d18da#diff-2c7147993fbd10a910bd1d5b1c93ef3c0c87f3cd170e07ab14a2a7e35598e454) to adopt too 👍 

Regarding the type adapters, I already have [a separated type adapter](https://github.com/beverloo/volunteer-manager/blob/main/app/lib/database/TemporalTypeAdapter.ts). You're right that it needs to be specified everywhere, but fortunately `ts-sql-codegen` makes this very easy.

The functions in [`DateTimeStringColumns`](https://github.com/beverloo/volunteer-manager/blob/main/app/lib/database/DateTimeStringColumns.ts) are used to avoid repetition in [`index.ts`](https://github.com/beverloo/volunteer-manager/blob/main/app/lib/database/index.ts#L63).

I actually tried a few approaches for that, including [something similar](https://github.com/beverloo/volunteer-manager/blob/c14fe5cd490ae92f5aa7571e102b644c1623f832/app/lib/database/Connection.ts#L56) to your suggestion (& something [far hackier that I've since deleted](https://github.com/beverloo/volunteer-manager/commit/bf4735ee924002a516f281ab9e14fedf7682f38f)), however columns are not always required - sometimes they can be `null`. That leads to quite a bit of duplication, and, more problematically for me, having to remember optionality of the columns myself in various places.

```typescript
class DBConnection extends MariaDBConnection<'DBConnection'> {
    zonedDateTimeAsString = this.buildFragmentWithArgs(
        this.arg<ZonedDateTime>('customLocalDateTime', 'ZonedDateTime', 'required')
    ).as((value) => {
        return this.fragmentWithType('string', 'required')
            .sql`date_format(${value}, "%Y-%m-%dT%TZ[UTC]")`
    });

    optionalZonedDateTimeAsString = this.buildFragmentWithArgs(
        this.arg<ZonedDateTime>('customLocalDateTime', 'ZonedDateTime', 'optional')
    ).as((value) => {
        return this.fragmentWithType('string', 'optional')
            .sql`date_format(${value}, "%Y-%m-%dT%TZ[UTC]")`
    });
}
```

Is there any way to avoid this?

I appreciate that you may break this at any time, and you absolutely should!

**Comment by juanluispaz** [comment]
Hi!

- Regarding the type adapter, if you use the separated adapter, remember to put it everywhere you define that type. Also, remember to put it in the `currentZonedDateTime` fragment. If you don't, when you query (if you do that) that value from the database to the backend, you can have issues.

- Regarding `DateTimeStringColumns`, be aware, there is something called [generated/computed columns](https://dev.mysql.com/doc/refman/8.0/en/create-table-generated-columns.html) [2](https://dev.mysql.com/doc/refman/8.0/en/create-table.html) [3](https://mariadb.com/kb/en/generated-columns/) [4](https://mariadb.com/kb/en/create-table/#generated-columns) that allows you to put those definitions inside the data model itself without consuming disk space (if you don't want), they are supported by ts-sql-query and by ts-sql-codegen. I'm not sure if that will be valuable in your case, but I wanted to point out there is that option.

- Regarding the fragment with an optional value, how you describe it is how we addressed it in the past. Still, I just added, for you, a new feature to build the overloaded fragment that returns required or optional depending on the provided arguments; with this feature, the code will look like:

```ts
class DBConnection extends MariaDBConnection<'DBConnection'> {

    zonedDateTimeAsString = this.buildFragmentWithMaybeOptionalArgs(
        this.arg<ZonedDateTime>('customLocalDateTime', 'ZonedDateTime', 'optional')
    ).as((value) => {
        return this.fragmentWithType('string', 'optional')
            .sql`date_format(${value}, "%Y-%m-%dT%TZ[UTC]")`
    });
}
```

I just released this new feature in ts-sql-query 1.62.0. See the [documentation](https://ts-sql-query.readthedocs.io/en/latest/queries/sql-fragments/#select-with-custom-reusable-sql-fragment-maybe-optional)

Let me know if that works for you and your feedback.

**Comment by beverloo** [comment]
Whoa—thank you so much! That's incredible, it works like a charm ([1](https://github.com/beverloo/volunteer-manager/commit/d36fa1794ebceab66bf80f8582da3f6f48567c2a#diff-2c7147993fbd10a910bd1d5b1c93ef3c0c87f3cd170e07ab14a2a7e35598e454), [2](https://github.com/beverloo/volunteer-manager/commit/e0675f0be5b96ce7690f86aa7859a54b608fab54#diff-2c7147993fbd10a910bd1d5b1c93ef3c0c87f3cd170e07ab14a2a7e35598e454)) and the `DateTimeStringColumns.ts` file no longer is a thing. All tests pass, & no errors were seen following the deploy.

I also had a look at generated columns, and it looks like MariaDB incorrectly infers nullability (it always assumes `null`) and [doesn't support setting it explicitly](https://jira.mariadb.org/browse/MDEV-10964), at least for `VIRTUAL` columns. I'll see if I can work around that 😃 

Thank you once again!!
